### ISSUE TITLE
IP and MAC address getting null

### ISSUE THREAD

**Issue by DashThink** [yavijava/yavijava#249]
Hi,

I am trying to get MAC address but getting null. And when i try to get IP address it returns IP6. I want IP4.
Can anybody help me here.
I am  trying to get sing below code: 

vm.getGuest().getIpAddress();
vm.getGuest().getNet()[0].getMacAddress();

Thanks

**Comment by michaelrice** [comment]
Do you have VMWare Tools installed on the guest os?

**Comment by DashThink** [comment]
No, Not installed..Is it necessary?

**Comment by michaelrice** [comment]
Yes, and the machine has to be powered on as well. You might want to check out the docs: http://www.yavijava.com/docs/vim.vm.GuestInfo.html

**Comment by DashThink** [comment]
I tried by installing vmware tool but same error

**Comment by michaelrice** [comment]
A few things going on here:

**Prerequisites.** You need VMware Tools running on the guest *and* the VM powered on. Without Tools, the guest info object is mostly empty and `getMacAddress()` returns null. You mentioned installing Tools — make sure the Tools service is actually running (not just installed) and that you wait long enough after power-on for Tools to initialize and report back to the host.

**Why `getIpAddress()` returns IPv6.** `vm.getGuest().getIpAddress()` returns the *primary* IP as reported by Tools. On modern guests that's often the IPv6 link-local or SLAAC address, not the IPv4 address you actually want.

To get IPv4 specifically, walk the per-NIC info and filter:

```java
GuestInfo guest = vm.getGuest();
if (guest == null || guest.getNet() == null) {
    return; // tools not running yet
}
for (GuestNicInfo nic : guest.getNet()) {
    String mac = nic.getMacAddress(); // null until Tools reports
    if (nic.getIpConfig() == null || nic.getIpConfig().getIpAddress() == null) {
        continue;
    }
    for (NetIpConfigInfoIpAddress addr : nic.getIpConfig().getIpAddress()) {
        // origin: manual, dhcp, linklayer, random, other
        // an IPv4 address is dotted-quad; IPv6 contains ':'
        if (!addr.getIpAddress().contains(":")) {
            System.out.println(mac + " -> " + addr.getIpAddress() + "/" + addr.getPrefixLength());
        }
    }
}
```

`NetIpConfigInfoIpAddress.origin` (a `NetIpConfigInfoIpAddressOrigin` enum) tells you whether the address came from DHCP, was statically assigned, link-layer, etc., if you need to filter further.

If MAC is still null after Tools is fully up and the VM has been running for a minute or two, that's a Tools-side reporting issue (often a permissions or network driver problem inside the guest), not a yavijava bug.

Closing — please reopen with details if Tools is confirmed running and you're still seeing nulls.
